cmd/soroban-rpc/internal/db/transaction.go

```diff
@@ -0,0 +1,336 @@
+	ledgerRange.LastLedger.Sequence = lcm1.LedgerSequence()
+	ledgerRange.LastLedger.CloseTime = lcm1.LedgerCloseTime()
+
+	if !rows.Next() {
+		return ledgerRange, errors.Wrap(rows.Err(), "couldn't query ledger range") // nil if no row err
+	} else if err := rows.Scan(&lcm2); err != nil {
+		return ledgerRange, errors.Wrap(err, "couldn't read ledger")
+	}
+
+	ledgerRange.LastLedger.Sequence = lcm2.LedgerSequence()
+	ledgerRange.LastLedger.CloseTime = lcm2.LedgerCloseTime()
+
+	txn.log.Debugf("Database ledger range: [%d, %d]",
+		ledgerRange.FirstLedger.Sequence, ledgerRange.LastLedger.Sequence)
+	return ledgerRange, nil
+}
+
+// GetTransaction conforms to the interface in
+// methods/get_transaction.go#NewGetTransactionHandler so that it can be used
+// directly against the RPC handler.
+//
+// Errors occur if there are issues with the DB connection or the XDR is
+// corrupted somehow. If the transaction is not found, io.EOF is returned.
+func (txn *transactionHandler) GetTransaction(ctx context.Context, hash xdr.Hash) (
+	Transaction, ledgerbucketwindow.LedgerRange, error,
+) {
+	start := time.Now()
+	tx := Transaction{}
+
+	ledgerRange, err := txn.GetLedgerRange(ctx)
+	if err != nil && err != ErrEmptyDB {
+		return tx, ledgerRange, err
+	}
+
+	lcm, ingestTx, err := txn.getTransactionByHash(ctx, hash)
+	if err != nil {
+		return tx, ledgerRange, err
+	}
+	tx, err = ParseTransaction(lcm, ingestTx)
+	if err != nil {
+		return tx, ledgerRange, err
+	}
+
+	txn.log.
+		WithField("txhash", hex.EncodeToString(hash[:])).
+		WithField("duration", time.Since(start)).
+		Debugf("Fetched and encoded transaction from ledger %d", lcm.LedgerSequence())
+
+	return tx, ledgerRange, nil
+}
+
+// getTransactionByHash actually performs the DB ops to cross-reference a
+// transaction hash with a particular set of ledger close meta and parses out
+// the relevant transaction efficiently by leveraging the `application_order` db
+// field.
+//
+// Note: Caller must do input sanitization on the hash.
+func (txn *transactionHandler) getTransactionByHash(ctx context.Context, hash xdr.Hash) (
+	xdr.LedgerCloseMeta, ingest.LedgerTransaction, error,
+) {
+	rowQ := sq.Select("t.application_order", "lcm.meta").
+		From(fmt.Sprintf("%s t", transactionTableName)).
+		Join(fmt.Sprintf("%s lcm ON (t.ledger_sequence = lcm.sequence)", ledgerCloseMetaTableName)).
+		Where(sq.Eq{"t.hash": []byte(hash[:])}).
+		Limit(1)
+
+	rows, err := txn.db.Query(ctx, rowQ)
+	if err != nil {
+		return xdr.LedgerCloseMeta{}, ingest.LedgerTransaction{},
+			errors.Wrapf(err, "failed row read for txhash %s", hash)
+	}
+	if !rows.Next() {
+		if ierr := rows.Err(); ierr != nil {
+			return xdr.LedgerCloseMeta{}, ingest.LedgerTransaction{},
+				errors.Wrapf(ierr, "failed to fetch txhash %s", hash)
+		}
+
+		// We use EOF to mark "this tx doesn't exist," since that isn't a real error.
+		return xdr.LedgerCloseMeta{}, ingest.LedgerTransaction{}, io.EOF
```

Comment:
I think it would be cleaner to return a boolean indicating that that transaction is present / not present rather than returning `io.EOF` as an indicator that the transaction is not present

Reply:
Four return values? :sweat_smile: how about `db.ErrNoTx`, instead? It overloads the term but actually kinda works lol

Reply:
sure, that works!

Reply:
opted for a custom error, instead :+1: